test(x/twap): add randomized geometric twap test on a balancer pool

p0mvn opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## What is the purpose of the change

This PR adds a geometric twap to now randomized test against a balancer pool. It creates 1000 pools with randomized supply and executes twap over random duration for these pools.

The result is asserted to be within "0.0001" relative to spot price.

This test serves as a sanity check that we do not encounter any internal panics in math calculations

## Testing and Verifying

This change added tests.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

stackman27 commented on x/twap/api_test.go:
```diff
@@ -837,3 +841,70 @@ func (s *TestSuite) TestGetArithmeticTwapToNow_ThreeAsset() {
+				ExitFee: sdk.ZeroDec(),
+			})
+
+			//
```
Were you going to add a comment here?

p0mvn replied:
Updated:

```
// We add 1ms to avoid always landing on the same block time
// In that case, the most recent spot price would be used
// instead of interpolation.
```

I think that answers your other question as well

stackman27 commented on x/twap/api_test.go:
```diff
@@ -837,3 +841,70 @@ func (s *TestSuite) TestGetArithmeticTwapToNow_ThreeAsset() {
+			})
+
+			//
+			oldTime := ctx.BlockTime().Add(1 * time.Millisecond)
```
Why can't oldTime just be the current blockTime? Why add 1 * time.Millisecond?

p0mvn replied:
https://github.com/osmosis-labs/osmosis/pull/3844#discussion_r1056018157

stackman27 commented on x/twap/api_test.go:
```diff
@@ -837,3 +841,70 @@ func (s *TestSuite) TestGetArithmeticTwapToNow_ThreeAsset() {
+			twap, err := app.TwapKeeper.GetGeometricTwapToNow(ctx, 1, denom0, denom1, oldTime)
+			s.Require().NoError(err)
+
+			osmomath.ErrTolerance{
```
Out of curiosity, why not check multiplicative tolerance too?

p0mvn replied:
Additive just gives a better idea of how many precision points there are here for all test cases. The goal of this test is just to see that twap is roughly close to spot price. So any reasonable tolerance would suffice.

stackman27 replied:
sounds good

ValarDragon replied:
Hrmm, I think the guarantee we want is actually the multiplicative tolerance though. Since spot prices can differ in orders of magnitude, this may be the wrong order of magnitude for checking many of the pools.

p0mvn replied:
Yeah but the largest spot price possible in this setup is `1_000_000_000 / 10_000  =  100_000` which should not cause an extremely high additive difference.

I think I will adjust the parameters to allow for even greater spot prices and switch to multiplicative tolerance

czarcas7ic commented on x/twap/api_test.go:
```diff
@@ -837,3 +841,75 @@ func (s *TestSuite) TestGetArithmeticTwapToNow_ThreeAsset() {
+			s.Require().NoError(err)
+
+			osmomath.ErrTolerance{
+				AdditiveTolerance: sdk.NewDecWithPrec(1, 4),
+			}.CompareBigDec(
+				osmomath.BigDecFromSDKDec(spotPrice),
+				osmomath.BigDecFromSDKDec(twap),
+			)
```
Do you expect the additive tolerance to go both ways? My guess is yes, but I just wanted to be sure.

p0mvn replied:
There are no guarantees on the direction of the tolerance because exp2 function does not guarantee it. However, log always rounds down.

So I would say it is more likely to be smaller than higher but I don't think we can claim that for all test cases

czarcas7ic reviewed (approved):
Despite me not having a ton of context on geo twap, the test was easy to follow. Nicely done 👍

p0mvn commented:
Addressed comments.

The initial supply range of valid values has been increased. Multiplicative tolerance is now used. Also, reduced the number of retries to 200 only to avoid increasing CI time by over 2 seconds
